### PR TITLE
Support migration from The Things Stack

### LINKED ISSUE #42: Add TTS Source
#### Summary

Add TTS source to support migration from TTS CE to C or other TTS->TTS migrations. See referenced thread here -> https://github.com/TheThingsIndustries/lorawan-stack-docs/issues/542#issuecomment-1087496671

#### What is already there? What do you see now?

Chirpstack

#### What is missing? What do you want to see?

TTS

#### How do you propose to implement this?

...

#### How do you propose to test this?

...

#### Can you do this yourself and submit a Pull Request?

no

### LINKED ISSUE #46: Support migration from The Things Stack
#### Summary

Support migration from TTS to TTS.

#### Why do we need this?

To ease migration between community and enterprise TTS.

#### What is already there? What do you see now?

The migration is not yet supported in this tool.

#### What is missing? What do you want to see?

The migration is not yet supported in this tool.

#### Environment

n/a

#### How do you propose to implement this?

...

#### How do you propose to test this?

...

#### Can you do this yourself and submit a Pull Request?

Yes.

### PR DESCRIPTION
**Pull request by happyRip**
#### Summary

Closes #42 

#### Changes

- Added `ttnv3` source 

#### Testing

Tested by creating a device on [eu1.cloud.thethings.network](https://eu1.cloud.thethings.network)
- created test device
- copied `ca.pem` generated in `lorawan-stack`
- set environmental variables
  - image
- ```bash
  $ go run cmd/ttn-lw-migrate/main.go devices eui-70b3d57ed005528c --source ttnv3 --verbose | jq
  ```
  - ![image](https://user-images.githubusercontent.com/73070465/189144860-686c6c1e-4001-48b5-b342-e1c90b3aa5fa.png)
- checked if downlinks got disabled for the exported device
  - ![image](https://user-images.githubusercontent.com/73070465/189144732-63be41c1-cf65-4370-afb3-76695ff9cfe6.png)
  - ![image](https://user-images.githubusercontent.com/73070465/189145172-099cef95-6311-48b8-b68d-aaf5fc208247.png)
    - ```json
      {
        "mac_settings": {
          "rx2_data_rate_index": 0,
          "rx2_frequency": "869525000",
          "schedule_downlinks": false
        }
      }
      ```

##### Regressions

n/a

#### Notes for Reviewers

@adriansmares 
- export section of each source is very similar -> possibly the `README.md` could be updated to improve it's format (separate export section, list of sources with just specific notes)
- `ttnv3` flags/variables could be improved a bit by me since every server has the same address there could be a `TTNV3_GRPC_ADDRESS` variable that sets the other flags by default, but can be overwritten if needed

Bonus:
- [`README.md` preview shortcut link](https://github.com/TheThingsNetwork/lorawan-stack-migrate/blob/feat/46-support-ttnv3-migration/README.md#the-things-stack)

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.

### CODE REVIEW

File: pkg/source/ttnv3/source.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+// RangeDevices implements the source.Source interface.
+func (s Source) RangeDevices(appID string, f func(source.Source, string) error) error {
+	is, err := api.Dial(s.ctx, s.config.identityServerGRPCAddress)
+	if err != nil {
+		return err
+	}
+	paths := ttnpb.RPCFieldMaskPaths["/ttn.lorawan.v3.EndDeviceRegistry/List"].Allowed
+	res, err := ttnpb.NewEndDeviceRegistryClient(is).List(
```
Comment (adriansmares) [inline pkg/source/ttnv3/source.go:181]: Note that you will need pagination here - `List` will not return all of the end devices, and you need to use the `limit`/`page`/`order`.

Comment (happyRip) [reply pkg/source/ttnv3/source.go]: Fixed in https://github.com/TheThingsNetwork/lorawan-stack-migrate/pull/47/commits/fdc52a0a996ba1edaee8682574a866e3f64a9020

File: pkg/source/ttnv3/auth.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	apiKey, err := flags.GetString(flagWithPrefix("app-access-key"))
+	if err != nil {
+		return err
+	}
+	if apiKey != "" {
+		logger.Debug("Using API key")
+		api.SetAuth("bearer", apiKey)
+		return nil
+	}
+	token := NewToken()
```
Comment (adriansmares) [inline pkg/source/ttnv3/auth.go:22]: Let's drop this mechanism completely. It's either API key or nothing else.

File: pkg/source/ttnv3/config.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	applicationServerGRPCAddress string
+	networkServerGRPCAddress     string
+
+	dryRun bool
+}
+
+func flagSet() *pflag.FlagSet {
+	flags := &pflag.FlagSet{}
+	flags.String(flagWithPrefix("app-id"), os.Getenv("TTNV3_APP_ID"), "TTS Application ID")
+	flags.String(flagWithPrefix("app-access-key"), os.Getenv("TTNV3_APP_ACCESS_KEY"), "TTS Application Access Key (with 'devices' permissions)")
```
Comment (adriansmares) [inline pkg/source/ttnv3/config.go:39]: Let's call this `app-api-key` and also `TTNV3_APP_API_KEY`.

File: pkg/source/ttnv3/config.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		return b
+	}
+
+	appID := stringFlag(flagWithPrefix("app-id"))
+	if appID == "" {
+		return nil, errNoAppID
+	}
+
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{}
+	if caPath := stringFlag(flagWithPrefix("ca-file")); caPath == "" {
```
Comment (adriansmares) [inline pkg/source/ttnv3/config.go:64]: Secure/insecure should be controllable by a flag (`--insecure`) and secure should be default. The CA field should be appended if provided, but can be missing.

Our production deployments use certificates that are signed by common CAs.

File: pkg/source/ttnv3/config.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		api.SetInsecure(true)
+		logger.Warn("Using insecure connection to API")
+	} else {
+		rand.Seed(time.Now().UnixNano())
```
Comment (adriansmares) [inline pkg/source/ttnv3/config.go:68]: This is not needed.

File: pkg/source/ttnv3/source.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+	res, err := s.getEndDevice(ids, nsPaths, asPaths, jsPaths, true)
+	if err != nil {
+		return nil, err
+	}
+	paths := ttnpb.AddFields(nsPaths, ttnpb.AddFields(asPaths, ttnpb.AddFields(jsPaths, "ids.dev_addr")...)...)
+	if err := dev.SetFields(res, paths...); err != nil {
+		return nil, err
+	}
+	// Clear all "{server}_address" fields from exported device
+	if err := dev.SetFields(&ttnpb.EndDevice{}, "application_server_address", "join_server_address", "network_server_address"); err != nil {
```
Comment (adriansmares) [inline pkg/source/ttnv3/source.go:73]: I think even using `nil` as first parameter is supported.

**Review by adriansmares — COMMENTED** [review]
Cool stuff.

File: pkg/source/ttnv3/config.go (hunk 5 above)
Comment (happyRip) [reply pkg/source/ttnv3/config.go:77]: The preview broke, but if it is related to `boolFlag` as I could see in the email it is now used to retrieve the `insecure` flag

Comment (adriansmares) [reply pkg/source/ttnv3/config.go:77]: The `rand.Seed` call in the secure branch.

File: pkg/source/ttnv3/source.go (hunk 6 above)
Comment (happyRip) [reply pkg/source/ttnv3/source.go:73]: Works indeed, fixed

**Comment by adriansmares** [comment]
> export section of each source is very similar -> possibly the `README.md` could be updated to improve it's format (separate export section, list of sources with just specific notes)

Can be done in a separate PR.

> `ttnv3` flags/variables could be improved a bit by me since every server has the same address there could be a `TTNV3_GRPC_ADDRESS` variable that sets the other flags by default, but can be overwritten if needed

The addresses are relevant for devices in `nam1`, `au1` etc., so no harm in having them explicit for now.

**Pull request merged**